File: product-app/src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import { Product } from '../models/product';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import * as signalR from '@aspnet/signalr';

@Injectable({
  providedIn: 'root'
})
export class ProductService {

  private _products: Array<Product>;

  private hubConnection: signalR.HubConnection;

  public get products(): Array<Product> {
    return this._products;
  }


  public set products(v: Product[]) {
    this._products = v;
  }


  addProduct(product: Product) {
    product.id = 0;
    this.http.post('http://localhost:5000/api/products', product, { observe: 'response' }).subscribe((resp) => {
      resp.ok ? console.log('SuccessPostRequest') : console.log(resp.body);
    });
  }

  removeProduct(id: number) {
    this.http.delete(`http://localhost:5000/api/products/${id}`, { observe: 'response' }).subscribe((resp) => {
      resp.ok ? console.log('SeccessDeleteRequest') : console.log(resp.body);
    });
  }

  editProduct(product: Product) {
    this.http.put(`http://localhost:5000/api/products/${product.id}`, product, {observe: 'response'}).subscribe((resp) => {
      resp.ok ? console.log('SeccessPutRequest') : console.log(resp.body);
    });
  }

  getProducts() {
    this.http.get('http://localhost:5000/api/products').subscribe((data: Product[]) => {
      this.products = data;
    });
  }

  constructor(private http: HttpClient) {
    this.getProducts();

    this.hubConnection = new signalR.HubConnectionBuilder().withUrl('http://localhost:5000/producthub').build();
    this.hubConnection
    .start()
    .then(() => {
      console.log('Connection started');
    })
    .catch(err => {
      console.log(err);
    });

    this.hubConnection.on('AddProduct', (prod: Product) => {
      this.products.push(prod);
    });

    this.hubConnection.on('EditProduct', (id: number, prod: Product) => {
      const product = this.products.find(p => p.id === id);
      product.name = prod.name;
      product.cost = prod.cost;
      product.category = prod.category;
      product.available = prod.available;
      product.count = prod.count;
      product.removeDate = prod.removeDate;
    });

    this.hubConnection.on('RemoveProduct', (id: number) => {
      this.products.splice(this.products.findIndex(p => p.id === id), 1);
    });
  }
}

File: ProductApp/ProductApp/Services/ProductService.cs
using Microsoft.AspNetCore.SignalR;
using ProductApp.Data;
using ProductApp.Models;
using ProductApp.Services.Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace ProductApp.Services {
   public class ProductService : IProductService {
      private readonly ProductsDbContext dbContext;
      private readonly IHubContext<ProductHub, IProductHub> hubContext;

      public ProductService(ProductsDbContext dbContext, IHubContext<ProductHub, IProductHub> hubContext) {
         this.dbContext = dbContext;
         this.hubContext = hubContext;
      }

      public async Task AddProduct(Product prod) {
         this.dbContext.Products.Add(prod);
         await this.dbContext.SaveChangesAsync();
         await this.hubContext.Clients.All.AddProduct(prod);
      }

      public async Task EditProduct(int id, Product prod) {
         var product = this.dbContext.Products.Find(id);
         product.Name = prod.Name;
         product.Available = prod.Available;
         product.Category = prod.Category;
         product.Cost = prod.Cost;
         product.Count = prod.Count;
         product.RemoveDate = prod.RemoveDate;
         this.dbContext.SaveChanges();
         await this.hubContext.Clients.All.EditProduct(id, prod);
      }

      public async Task<IEnumerable<Product>> GetProducts() {
         return this.dbContext.Products.ToList();
      }

      public async Task RemoveProduct(int id) {
         this.dbContext.Products.Remove(
             this.dbContext.Products.Find(id)
         );
         await this.dbContext.SaveChangesAsync();
         await this.hubContext.Clients.All.RemoveProduct(id);
      }
   }
}
